### PR TITLE
Make contacts and organizations cloneable.

### PR DESCRIPTION
**Pull request by vanzegbroecke**
We want to be able to clone Organizations and Contacts. Use case: MemberChangedEvent handlers which update the member. We can't update the newEntity because otherwise other handlers will not get the original newEntity.

Depends on: https://github.com/VirtoCommerce/vc-platform/pull/1524

### CODE REVIEW

File: VirtoCommerce.Domain/Customer/Model/Organization.cs [hunk 1]
```diff
@@ -1,11 +1,50 @@
...
         public string Description { get; set; }
         public string BusinessCategory { get; set; }
         public string OwnerId { get; set; }
         public string ParentId { get; set; }
 
+        public virtual object Clone()
+        {
+            var clone = (Organization) MemberwiseClone();
+
+            clone.Addresses = Addresses
```
Comment (yecli) [inline VirtoCommerce.Domain/Customer/Model/Organization.cs]: I think we need to add support ICloneable to Member (and probably to all its descendants). MemberwiseClone and cloning its collection fields should be done inside base Member class, descendants should call base implementation + add own info.

File: VirtoCommerce.Domain/Customer/Model/Contact.cs [hunk 2]
```diff
@@ -27,8 +28,24 @@ public class Contact : Member, IHasSecurityAccounts
...
 
         #endregion
+
+        public virtual object Clone()
+        {
+            var clone = (Contact) MemberwiseClone();
+
+            if (Organizations != null)
+            {
+                clone.Organizations = new List<string>(Organizations);
```
Comment (yecli) [inline VirtoCommerce.Domain/Customer/Model/Contact.cs]: We need to make Organizations list copy, 
<code>new List<string>(Organizations)</code>
creates list referencing the same strings.
.ToList() could help.

File: VirtoCommerce.Domain/Customer/Model/Organization.cs [hunk 3]
```diff
@@ -1,11 +1,50 @@
...
+                ?.Select(x => (Address) x.Clone())
+                .ToList();
+
+            if (Phones != null)
+            {
+                clone.Phones = new List<string>(Phones);
```
Comment (yecli) [inline VirtoCommerce.Domain/Customer/Model/Organization.cs]: Better use ToList() here.

File: VirtoCommerce.Domain/Customer/Model/Organization.cs [hunk 4]
```diff
@@ -1,11 +1,50 @@
...
+            }
+
+            if (Emails != null)
+            {
+                clone.Emails = new List<string>(Emails);
```
Comment (yecli) [inline VirtoCommerce.Domain/Customer/Model/Organization.cs]: Better use ToList() here.

File: VirtoCommerce.Domain/Customer/Model/Organization.cs [hunk 5]
```diff
@@ -1,11 +1,50 @@
...
+            }
+
+            clone.Notes = Notes
+                ?.Select(x => (Note) x.Clone())
+                .ToList();
+
+            if (Groups != null)
+            {
+                clone.Groups = new List<string>(Groups);
```
Comment (yecli) [inline VirtoCommerce.Domain/Customer/Model/Organization.cs]: Better use ToList() here.

File: VirtoCommerce.Domain/Customer/Model/Contact.cs (hunk 2 above)
Comment (vanzegbroecke) [reply VirtoCommerce.Domain/Customer/Model/Contact.cs]: Cloning strings isn't really necessary is it? https://referencesource.microsoft.com/#mscorlib/system/string.cs

File: VirtoCommerce.Domain/Customer/Model/Organization.cs (hunk 3 above)
Comment (vanzegbroecke) [reply VirtoCommerce.Domain/Customer/Model/Organization.cs]: Coding preference? The implementation of ToList does the same thing.

File: VirtoCommerce.Domain/Customer/Model/Organization.cs (hunk 1 above)
Comment (vanzegbroecke) [reply VirtoCommerce.Domain/Customer/Model/Organization.cs]: I completely agree. See commit 4a76c3ff26a45968f1b4c3aa9acd1b3332f461b0.

File: VirtoCommerce.Domain/Customer/Model/Contact.cs (hunk 2 above)
Comment (yecli) [reply VirtoCommerce.Domain/Customer/Model/Contact.cs]: OK, let's leave it that way.

File: VirtoCommerce.Domain/Customer/Model/Organization.cs (hunk 3 above)
Comment (yecli) [reply VirtoCommerce.Domain/Customer/Model/Organization.cs]: Let's leave it then

File: VirtoCommerce.Domain/Customer/Model/Contact.cs (hunk 2 above)
Comment (artem-dudarev) [reply VirtoCommerce.Domain/Customer/Model/Contact.cs]: Actually `.ToList()` is much better. It is shorter and it alows to remove `
if (Organizations != null)`:
```csharp
clone.Organizations = Organizations?.ToList();
```

Comment (vanzegbroecke) [reply VirtoCommerce.Domain/Customer/Model/Contact.cs]: Has already been refactored: https://github.com/VirtoCommerce/vc-module-core/pull/111/commits/4a76c3ff26a45968f1b4c3aa9acd1b3332f461b0

**Comment by vanzegbroecke** [comment]
@tatarincev Could someone please complete this PR?

**Pull request merged**